types/lazy/deferred.go

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package lazy
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"tailscale.com/types/ptr"
+)
+
+// DeferredInit allows one or more funcs to be deferred
+// until [DeferredInit.Do] is called for the first time.
+//
+// DeferredInit is safe for concurrent use.
+type DeferredInit struct {
+	DeferredFuncs
+}
+
+// DeferredFuncs allows one or more funcs to be deferred
+// until the owner's [DeferredInit.Do] method is called
+// for the first time.
+//
+// DeferredFuncs is safe for concurrent use.
+type DeferredFuncs struct {
+	m     sync.Mutex
+	funcs []func() error
+
+	// err is either:
+	//    * nil, if deferred init has not yet been completed
+	//    * nilErrPtr, if initialization completed successfully
+	//    * non-nil and not nilErrPtr, if there was an error
+	//
+	// It is an atomic.Pointer so it can be read without m held.
+	err atomic.Pointer[error]
+}
+
+// Defer adds a function to be called when [DeferredInit.Do]
+// is called for the first time. It returns true on success,
+// or false if [DeferredInit.Do] has already been called.
+func (d *DeferredFuncs) Defer(f func() error) bool {
+	d.m.Lock()
+	defer d.m.Unlock()
+	if d.err.Load() != nil {
+		return false
+	}
+	d.funcs = append(d.funcs, f)
+	return true
+}
+
+// MustDefer is like [DeferredFuncs.Defer], but panics
+// if [DeferredInit.Do] has already been called.
+func (d *DeferredFuncs) MustDefer(f func() error) {
+	if !d.Defer(f) {
+		panic("deferred init already completed")
+	}
+}
+
+// Do calls previously deferred init functions if it is being called
+// for the first time on this instance of [DeferredInit].
+// It stops and returns an error if any init function returns an error.
+//
+// It is safe for concurrent use, and the deferred init is guaranteed
+// to have been completed, either successfully or with an error,
+// when Do() returns.
+func (d *DeferredInit) Do() error {
+	err := d.err.Load()
+	if err == nil {
+		err = d.doSlow()
+	}
```

Comment:
Rather than an an `atomic.Pointer`, it might be easier to read with a `sync.Once` that sets an `err` field that you can return directly (no atomic or mutex needed, since it's guaranteed to be immutable after the `sync.Once` runs). It's implemented roughtly the same (an atomic uint32 + a mutex).

Reply:
The mutex+atomic approach was chosen mainly for two reasons:
1. Deferring functions (appending to the `funcs` slice) needs to be concurrency-safe, so we needed a mutex anyway;
2. Ideally, `(*DeferredInit).Defer` should fail (and report it to the caller) if `(*DeferredInit).Do()` has already been called. Since `sync.Once` doesn't allow checking if it has already been called, we need to manage the state ourselves.

Therefore, it would either be a `bool` + a mutex + a `sync.Once`, or just an atomic+mutex.